Sign in teachers using a one time password

thomasleese opened the pull request:
This implements a "one-time password" authentication mechanism for teachers based on the same code used in https://github.com/DFE-Digital/refer-serious-misconduct.

Depends on #815 

[Trello Card](https://trello.com/c/nLrL1b0P/1235-update-auth-to-use-otp-instead-of-not-so-magic-links)

## Screenshots

<img width="674" alt="Screenshot 2022-12-09 at 11 53 37" src="https://user-images.githubusercontent.com/510498/206696903-af6c62d5-d804-446b-aba2-37caef459d13.png">

<img width="683" alt="Screenshot 2022-12-10 at 22 29 57" src="https://user-images.githubusercontent.com/510498/206878292-9454b140-bb02-4d6b-bf84-c7f3672e3804.png">

<img width="566" alt="Screenshot 2022-12-10 at 22 30 46" src="https://user-images.githubusercontent.com/510498/206878295-db4cc7b2-db84-4896-a57b-b30d32e375e0.png">

<img width="548" alt="Screenshot 2022-12-10 at 22 32 25" src="https://user-images.githubusercontent.com/510498/206878297-6c87f047-eba7-44d6-bb81-5b801a38a9d4.png">

gpeng commented on app/forms/teachers/otp_form.rb:
```diff
@@ -0,0 +1,52 @@
+
+class Teachers::OtpForm
+  MAX_GUESSES = 5
+  EXPIRY_IN_MINUTES = 30.minutes
```
The email says 15 minutes

gpeng commented on app/lib/devise/strategies/otp_authenticatable.rb:
```diff
@@ -0,0 +1,53 @@
+
+  def authenticate!
+    resource = mapping.to.find_by(email:)
+    unless resource && resource.secret_key.present?
```
Could we extract these predicates here and on 27 so we get:

```
fail!(:otp_invalid) and return if secret_key_absent? || otp_invalid?
```

gpeng commented on app/views/teachers/mailer/otp.text.erb:
```diff
@@ -0,0 +1,11 @@
+
+<%= @otp %>
+
+The code will expire after 15 minutes.
```
15 minutes or 30?

gpeng reviewed (changes requested):
Looks good and works really well. So much better than those horrific links.

I spotted a couple of bits of content that need updating.

<img width="836" alt="Screenshot 2022-12-09 at 13 44 21" src="https://user-images.githubusercontent.com/5216/206718650-4874b97c-41ba-412d-acad-27a577491432.png">

<img width="721" alt="Screenshot 2022-12-09 at 13 41 37" src="https://user-images.githubusercontent.com/5216/206718592-726cbc89-bf70-44dd-90a9-a4962ede44fd.png">

There's a validation message missing too. Not sure we even want to validate fort that here 🤷 

<img width="753" alt="Screenshot 2022-12-09 at 13 42 43" src="https://user-images.githubusercontent.com/5216/206718821-14a4561b-9cbc-4980-83b5-f937e83154d4.png">

I also got a 500 page when I tried to enter a code that I'd already used. I didn't test the expiry as yet but reuse definitely errors.

gpeng commented:
Just spotted you've addressed some of that content stuff in the next PR. I'll check again on that one when this is merged.

thomasleese commented:
Thanks @gpeng I've addressed those comments.

> I also got a 500 page when I tried to enter a code that I'd already used. I didn't test the expiry as yet but reuse definitely errors.

Do you know if you took any other steps? I've got a test for this scenario and can't reproduce the problem myself: https://github.com/DFE-Digital/apply-for-qualified-teacher-status/blob/5ce576166a198a124d23564cac447e0ce1026545/spec/system/teacher_interface/authentication_spec.rb#L94-L107

> Do we need a 'resend the code' link on the 'enter your code' page.

It might be a good idea, would be interested to see what the UCD team think. There is a back link that takes them to the sign in page again, and the failure page has a link allowing them to try again.

> Also I wonder if we should include the link back to that page in the email although I realise that will get us back into horrible link land.

Yeah, I'm not sure about this one, I'm happy to add it, but I suppose then users might think it's a link they have to click to sign in.

gpeng commented:
> Do you know if you took any other steps? I've got a test for this scenario and can't reproduce the problem myself:

I can't recreate it now either 🤷
